### PR TITLE
test(core): SchemaMigrationFailed integration test (refs #983)

### LINKED ISSUE #983: [PLAN] test(core): SchemaMigrationFailed integration test for PluginLoader step 11 (iterate #231 + #221)
### Domain

domain:core

### Phase

phase:mvp

### Story Points (Fibonacci, > 8 must split)

pts:2

### User Stories Satisfied

Hot-reload swap user-stories. Full coverage of plugin-abi.md §"Loader Sequence" step 11 (migration failure → SchemaMigrationFailed).

### Parent (Epic or Sub-Epic)

#5

### Scope

Add an integration test (or injectable-failure unit test) that exercises the SchemaMigrationFailed error path of `PluginLoaderRegistry::migrate_components` (plugin-abi.md §"Failure Modes" step 11).

Options:
1. Inject a mock migration-table resolver (via function pointer or template) that returns `std::unexpected(core::Error::SchemaMigrationFailed)` to simulate a migration step failure.
2. Use the real migration table walker once plan #221 (glibre-foryc migration dispatcher) lands and construct a fixture with an intentionally broken migration function.

Out of scope: the real per-type migration table walk (that is #221).

### Unit Test Plan

- core/plugin_loader_registry: migrate_components_returns_failed_on_broken_migration_step

### Definition of Done

```yaml
- pr_merged_closes_self: true
- unit_test_named: migrate_components_returns_failed_on_broken_migration_step
- workflow_passed: ci.yml
```

### Depends on

- #221 (glibre-foryc migration dispatcher — needed for the real per-type migration path)

### Agent Execution Notes

Deferred from plan #231 per MED-3 round-1 review finding on PR #979 (step 11 zero coverage). Cite this issue in any PR that addresses it.

### PR DESCRIPTION
**Pull request by cjhowe-us**
## Summary

- Adds a `MigrationStepFn` injection seam (second overload of `migrate_components`) so the loader step-11 failure path can be exercised without the real per-type migration table walk from plan #221.
- The new overload accepts a `Result<void>(*)() noexcept` function pointer that is invoked only when `from_version != to_version`; existing production overload is unchanged.
- Adds the named Catch2 test `migrate_components_returns_failed_on_broken_migration_step` in `tests/core/plugin_loader_registry/` that injects a mock step returning `std::unexpected(SchemaMigrationFailed)` and asserts the error propagates correctly per plugin-abi.md §"Failure Modes" step 11.

## Files changed

- `core/include/glibre/core/plugin_loader_actions.hpp` — `MigrationStepFn` type alias + injectable overload declaration
- `core/src/plugin_loader_actions.cpp` — injectable overload implementation
- `tests/core/plugin_loader_registry/plugin_loader_registry_test.cpp` — new test case + `plugin_loader_actions.hpp` include
- `tests/core/plugin_loader_registry/CMakeLists.txt` — updated comment to reference plan #983

## Test plan

- [x] `migrate_components_returns_failed_on_broken_migration_step` passes (3 assertions in 1 test case)
- [x] All 171 existing tests continue to pass (no regressions)
- [x] `migrate_components_no_op_when_versions_equal` (the stub smoke test from plan #231) still passes — production overload is unchanged

Closes #983

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: tests/core/plugin_loader_registry/plugin_loader_registry_test.cpp [hunk 1]
```diff
@@ -551,3 +556,46 @@ TEST_CASE("name_unique_rejects_same_name_different_path", "[core][plugin_loader_
     REQUIRE(core_err2 != nullptr);
     CHECK(*core_err2 == glibre::core::Error::PluginNameCollision);
```
Comment (cjhowe-us) [inline tests/core/plugin_loader_registry/plugin_loader_registry_test.cpp:552]: severity:MED location:tests/core/plugin_loader_registry/plugin_loader_registry_test.cpp:557 problem:new migrate_components failure-path test placed in plugin_loader_registry/ but the existing migrate_components_no_op_when_versions_equal smoke test (plan #231) lives in tests/core/plugin_loader_register/. The same SUT (plugin_loader_actions.hpp free functions) is now exercised across two separate test binaries — cohesion drift. The plugin_loader_register/ directory is the established home for plugin_loader_actions.* tests (call_register, rebuild_schedule, migrate_components per its file header). plugin_loader_registry/ is for PluginLoaderRegistry gate tests — different SUT, different responsibility. fix:move the new TEST_CASE (and the plugin_loader_actions.hpp include) into tests/core/plugin_loader_register/plugin_loader_register_test.cpp; revert the include + test from plugin_loader_registry_test.cpp; revert the CMakeLists.txt named-test comment changes in plugin_loader_registry/ and add the new named test to plugin_loader_register/CMakeLists.txt instead.

File: core/src/plugin_loader_actions.cpp [hunk 2]
```diff
@@ -217,6 +232,41 @@ Result<void> migrate_components(
...
+    // step_fn is not called: callers may pass a valid mock without it running.
+    if (from_version == to_version) {
+        return {};
+    }
+
+    // A non-null step_fn is required when from != to.
+    // A null step_fn indicates a caller logic error (the caller must supply
+    // the migration step when versions differ); map to SchemaMigrationFailed
+    // so the caller's error handling path sees the same arm regardless of
+    // whether the step was null or the step itself failed.
```
Comment (cjhowe-us) [inline core/src/plugin_loader_actions.cpp]: severity:MED location:core/src/plugin_loader_actions.cpp:248 problem:the null-step_fn branch maps a caller-logic error (null function pointer when from!=to) to core::Error::SchemaMigrationFailed. plugin-abi.md §"Failure Modes" row 11 reads "migration step failed or chain missing" — "chain missing" arguably covers null, but the header docstring at plugin_loader_actions.hpp line 203 says "Must be non-null" (a precondition). The implementation diverges from the header in two ways: (a) it accepts the null pointer at runtime instead of treating it as a precondition violation, and (b) the new code path is added but is not exercised by any test in this PR — only the "step returns unexpected" arm is covered. fix:either (i) tighten the header to document that null-step_fn maps to SchemaMigrationFailed (drop the "Must be non-null" precondition) AND add a second SECTION/test asserting that arm; or (ii) drop the runtime null check and document it as a debug-build assert/UB precondition (consistent with the rest of the codebase, e.g. validate_all preconditions). Pick one consistently — the current state is a hidden untested branch.

File: tests/core/plugin_loader_registry/plugin_loader_registry_test.cpp [hunk 3]
```diff
@@ -551,3 +556,46 @@ TEST_CASE("name_unique_rejects_same_name_different_path", "[core][plugin_loader_
...
+    // The lambda has no captures so the conversion is valid under C++11+.
+    glibre::core::MigrationStepFn broken_step = []() noexcept -> glibre::Result<void> {
+        return std::unexpected(glibre::Error{glibre::core::Error::SchemaMigrationFailed});
+    };
+
+    // from_version (1) != to_version (2): a real schema bump — step_fn is
+    // invoked and its failure propagated to the caller.
+    auto result = glibre::core::migrate_components(1u, 2u, broken_step);
+
+    REQUIRE(!result);
```
Comment (cjhowe-us) [inline tests/core/plugin_loader_registry/plugin_loader_registry_test.cpp]: severity:LOW location:tests/core/plugin_loader_registry/plugin_loader_registry_test.cpp:597 problem:the test only asserts the SchemaMigrationFailed arm via REQUIRE+CHECK on the error code; it does not verify (a) that step_fn is actually invoked when from!=to, nor (b) that step_fn is NOT invoked when from==to (the identity-case contract documented in plugin_loader_actions.hpp lines 174-177 and the cpp identity-case branch). A captured invocation counter on the lambda would catch a regression where the production overload accidentally short-circuits or a future refactor swaps the order of the from==to check and the step_fn invocation. fix:either add a second TEST_CASE migrate_components_does_not_call_step_fn_when_versions_equal that asserts step_fn was not called via a static-counter helper, or (cheaper) augment the existing test with from==to + same broken_step and assert result.has_value() — proving the no-op short-circuit really fires before step_fn runs.

**Review by cjhowe-us — COMMENTED** [review]
round:1 reviewer:go-review verdict:COMMENT
findings:
  HIGH:0
  MED:2
  LOW:1
carryover:0
notes:Coverage-and-correctness pass. CI is green on latest commit (9309827, all required checks pass; e2e-core/e2e-shader skipping is expected for a core-only test PR). PR body contains "Closes #983"; Conv-Commit subject `test(core):` is correct; DoD on #983 is satisfied (pr_merged_closes_self via Closes-keyword, unit_test_named present, workflow_passed currently green). The injection seam is sound: production overload `migrate_components(from, to)` is untouched, the new overload is additive, MigrationStepFn is a plain function pointer (no eastl::function bloat — good cohesion with PHILOSOPHY §11), the noexcept qualifier matches -fno-exceptions per error-model.md Decision 3, the captureless-lambda → function-pointer conversion in the test is valid C++, and the test asserts both !result and *core_err == SchemaMigrationFailed as required. Two MED items: (1) test placement — the new TEST_CASE landed in tests/core/plugin_loader_registry/ but the existing migrate_components_no_op_when_versions_equal smoke test (plan #231) lives in tests/core/plugin_loader_register/; the same SUT (plugin_loader_actions.hpp free functions) is now split across two test binaries, breaking the directory-level SRP that plugin_loader_register/ is for actions and plugin_loader_registry/ is for the registry gates; (2) the new null-step_fn branch in the cpp is a code path that is not exercised by any test in this PR and conflicts with the header's "Must be non-null" precondition wording — pick precondition-OR-error-arm and document/test it consistently. One LOW: the test does not verify (a) that step_fn IS invoked on from!=to or (b) that step_fn is NOT invoked on from==to — an invocation counter would harden against future refactors. None of these block merge for round 1; recommend impl-respond to address them before round 2.

File: tests/core/plugin_loader_registry/plugin_loader_registry_test.cpp (hunk 1 above)
Comment (cjhowe-us) [reply tests/core/plugin_loader_registry/plugin_loader_registry_test.cpp:552]: responder:go-impl-respond round:1 decision:ADDRESSED commit:9e82891 reasoning:Moved migrate_components_returns_failed_on_broken_migration_step and its plugin_loader_actions.hpp include from tests/core/plugin_loader_registry/ to tests/core/plugin_loader_register/ — the established home for all plugin_loader_actions.* tests (call_register, rebuild_schedule, migrate_components). Reverted registry/CMakeLists.txt to restore #981 references and removed #983 from its named-test list. Added #983 entry + the new test case name to register/CMakeLists.txt. All plan #230 registry tests still pass; all plan #231/#983 register tests pass.

File: core/src/plugin_loader_actions.cpp (hunk 2 above)
Comment (cjhowe-us) [reply core/src/plugin_loader_actions.cpp]: responder:go-impl-respond round:1 decision:ADDRESSED commit:9e82891 reasoning:Chose path (a) — tighten the impl. Removed the runtime null-check branch from migrate_components() and replaced it with an #ifndef NDEBUG / assert() block, consistent with the codebase pattern in hot_reload_validate (plugin_loader_actions.cpp lines 350–361). The header @param step_fn docstring 'Must be non-null' is preserved unchanged as the precondition contract. There is now no untested runtime branch: the only code path executed when from!=to is the step_fn() call, which is exercised by the new test.

File: tests/core/plugin_loader_registry/plugin_loader_registry_test.cpp (hunk 3 above)
Comment (cjhowe-us) [reply tests/core/plugin_loader_registry/plugin_loader_registry_test.cpp]: responder:go-impl-respond round:1 decision:ADDRESSED commit:9e82891 reasoning:Added two SECTIONs to the moved test. Section A (from!=to): uses a TU-static counter (g_broken_step_call_count) incremented by a file-scope broken_step_fn() — counter is checked == 1 after the call, verifying step_fn was invoked exactly once before the error is propagated. Section B (from==to): counter reset to 0, call with from==to==3, counter checked == 0 afterwards — proving the identity short-circuit fires before step_fn invocation, and result.has_value() is asserted. Both sections use the same broken_step_fn so any regression that swaps the guard order causes the counter assertion to fail.

File: core/include/glibre/core/plugin_loader_actions.hpp [hunk 4]
```diff
@@ -149,16 +149,42 @@ namespace glibre::core {
...
 //
 // For each persistent component whose schema bumped versions, runs the
 // per-type deserialize<T> migration path against the pre-swap snapshot
 // (fory-codegen.md §"Migration Mechanic").
 //
-// MVP STUB: Returns success unconditionally when from_version == to_version
-// (no migration needed) and also when from_version != to_version (real
-// migration deferred — see TODO below).
+// Two overloads:
+//
```
Comment (cjhowe-us) [inline core/include/glibre/core/plugin_loader_actions.hpp:175]: severity:LOW location:core/include/glibre/core/plugin_loader_actions.hpp:172-176 problem:Header docstring for overload (1) reads "Uses the default no-op step function (always returns success)." That wording suggests overload (1) internally threads a built-in default step_fn through the same mechanism as overload (2). The actual implementation in plugin_loader_actions.cpp:209-218 takes no MigrationStepFn at all — it just returns `{}` directly. The docstring/impl mismatch is harmless functionally but creates a maintenance trap: a future reader refactoring the seam might assume overload (1) is `migrate_components(from, to, &default_noop_step)` and try to deduplicate by deleting it. fix:reword to e.g. "Returns success unconditionally in the MVP stub; no step_fn parameter — the real plan #221 implementation will walk a per-type migration table internally rather than accepting an injected step." Keeps the contract honest about the difference between the two overloads' shapes.

File: tests/core/plugin_loader_register/plugin_loader_register_test.cpp [hunk 5]
```diff
@@ -428,6 +428,87 @@ TEST_CASE("plugin_loader_stamps_allocator_handle_with_plugin_tag", "[core][regis
...
+// ===========================================================================
+
+// TU-local invocation counter for the broken_step lambda (accessed via a
+// file-scope helper so it can be used from a non-capturing function pointer).
+// Reset to 0 before each SECTION to avoid cross-section interference.
+namespace {
+static int g_broken_step_call_count =
+    0;  // NOLINT(cppcoreguidelines-avoid-non-const-global-variables)
+
+glibre::Result<void> broken_step_fn() noexcept {
```
Comment (cjhowe-us) [inline tests/core/plugin_loader_register/plugin_loader_register_test.cpp:466]: severity:LOW location:tests/core/plugin_loader_register/plugin_loader_register_test.cpp:464-470 problem:`static int g_broken_step_call_count = 0;` inside an anonymous namespace combines two redundant internal-linkage mechanisms — the anonymous namespace already gives internal linkage (C++11+), so the `static` storage-class specifier is redundant for linkage purposes (it does still control storage duration, but namespace-scope variables already have static storage duration). The `// NOLINT(cppcoreguidelines-avoid-non-const-global-variables)` is fine. Same pattern applies to broken_step_fn — it's already TU-local via the anonymous namespace. Minor seam-quality nit: the codebase elsewhere (e.g. as_core_error helper at the top of this same test file) uses anonymous-namespace-without-static. fix:drop the redundant `static` keyword for consistency with the rest of the file: `int g_broken_step_call_count = 0;`. Pure stylistic; not functional.

**Review by cjhowe-us — COMMENTED** [review]
round:2 reviewer:go-review verdict:COMMENT
findings:
  HIGH:0
  MED:0
  LOW:2
carryover:0
notes:Cohesion + SRP + seam-quality pass on commit ee910d8 (CI green: build pass 3m55s, lint/dep-check/spec-check/automated-review all pass; e2e-core/e2e-shader skipping is expected for a core-only test PR). All three R1 findings (MED-1 test placement, MED-2 null-step_fn untested branch, LOW-1 invocation assertions) are correctly addressed in 9e82891+ee910d8: test moved to tests/core/plugin_loader_register/ where every plugin_loader_actions.* test now lives (cohesion clean — same TU as migrate_components_no_op_when_versions_equal, call_register, rebuild_schedule); runtime null branch replaced by NDEBUG-gated assert that mirrors the hot_reload_validate precondition pattern in the same file (no untested production branch); the test now has two SECTIONs with a TU-local invocation counter (`g_broken_step_call_count` reset per section) that proves step_fn is invoked exactly once when from!=to and zero times when from==to — guards against any future refactor that swaps the order of the identity short-circuit and step_fn invocation. Public-API surface review: the new MigrationStepFn type + second overload are purely additive; production overload (`migrate_components(from, to)`) is byte-for-byte unchanged. The test-injection seam co-located in plugin_loader_actions.hpp follows the established convention (call_register's two-overload split for the same reason — production-with-stamping vs. unit-test-with-hand-crafted-context). Re-derives from glibre primitives (Result<void>, glibre::Error, core::Error::SchemaMigrationFailed); plain function pointer chosen over eastl::function to avoid heap+vtable per PHILOSOPHY §11; noexcept qualifier matches -fno-exceptions per error-model.md §Decision 3. Cross-test bleed risk on the static counter is low — no other test in the TU calls broken_step_fn. Two new LOWs surfaced in this round: (1) overload (1)'s docstring claims it "uses the default no-op step function" but the actual impl takes no step_fn parameter at all — minor docstring/impl mismatch that could mislead a future refactor toward an unsound dedup; (2) `static int g_broken_step_call_count` inside an anonymous namespace double-specifies internal linkage — drop the `static` keyword for consistency with the as_core_error helper higher in the same file. Both are pure quality polish and do not block merge. Recommend impl-respond address them before round 3 if convenient; if not, defer to round 3 closure.

File: core/include/glibre/core/plugin_loader_actions.hpp (hunk 4 above)
Comment (cjhowe-us) [reply core/include/glibre/core/plugin_loader_actions.hpp:175]: responder:go-impl-respond round:2 decision:ADDRESSED commit:1a449f1 reasoning:Rewrote overload (1) docstring to accurately describe its implementation: 'Takes no step_fn parameter; returns success unconditionally in the MVP stub. The plan #221 implementation will walk a per-type migration table internally rather than accepting an injected step. For tests that need an injection seam, use overload (2).' Removes the false implication that overload (1) threads a built-in step through overload (2).

File: tests/core/plugin_loader_register/plugin_loader_register_test.cpp (hunk 5 above)
Comment (cjhowe-us) [reply tests/core/plugin_loader_register/plugin_loader_register_test.cpp:466]: responder:go-impl-respond round:2 decision:ADDRESSED commit:1a449f1 reasoning:Dropped the redundant 'static' keyword from g_broken_step_call_count. Anonymous namespace already provides internal linkage and static storage duration; the explicit 'static' was double-specifying linkage, inconsistent with as_core_error at the top of the same TU. Now reads 'int g_broken_step_call_count = 0;' inside the anon-ns.

**Review by cjhowe-us — COMMENTED** [review]
round:3 reviewer:go-review verdict:APPROVE
findings:
  HIGH:0
  MED:0
  LOW:1
carryover:0
notes:Topology + spec-alignment + polish pass on commit cad0ed9 (CI green: build 5m29s, lint/dep-check/spec-check/automated-review all pass; e2e-core/e2e-shader skipping expected for core-only test PR). All R1 (3) and R2 (2) findings correctly resolved. Topology: #5 (sub-epic) -> #983 (plan) parenting is correct; dependencies #221 (foryc dispatcher, open) + #231 (PluginLoader register+schedule+migrate, open) both exist; #983 was unblocked-by-design via Option 1 (mock injection) per its own Scope, so the dod:failed state on #221/#231 does not block this leaf. Spec alignment verified directly against reviews/decisions/plugin-abi.md: Loader Sequence (line 231) and Failure Modes -> core::Error table (line 368, row 11: SchemaMigrationFailed) both present and cited correctly in the test header docstring and the impl comment. Polish: overload (1) docstring is now honest about the paramless MVP stub and forward-references plan #221 + overload (2); overload (2) docstring + impl + test are consistent on the from==to short-circuit before step_fn invocation; the design note in plugin_loader_actions.cpp explaining the plain-function-pointer choice (vs. eastl::function heap+vtable) ties back to PHILOSOPHY 11 cleanly. DoD on #983 fully satisfied: pr_merged_closes_self (Closes #983 in PR body), unit_test_named (migrate_components_returns_failed_on_broken_migration_step), workflow_passed (ci.yml green on cad0ed9). One new LOW: issue #983 body declares pts:2 but the label is missing -- pure issue-hygiene drift, non-blocking, posted as issue-level comment. No HIGH/MED concerns; recommend merge.
